foundation/foundation-upgrade/src/main/java/com/boozallen/aissemble/upgrade/migration/v1_11_0/MlPipelineDockerPomMigration.java

```diff
@@ -0,0 +1,139 @@
+import org.apache.maven.project.MavenProject;
+import org.codehaus.plexus.util.xml.Xpp3Dom;
+import org.technologybrewery.baton.BatonException;
+import org.technologybrewery.baton.util.pom.PomHelper;
+import org.technologybrewery.baton.util.pom.PomModifications;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.List;
+
+import static org.apache.commons.lang3.StringUtils.repeat;
+import static org.technologybrewery.baton.util.pom.LocationAwareMavenReader.END;
+
+/**
+ * This migration updates machine learning pipeline docker POM to include step artifact dependency for resolving the build cache dependency check issue.
+ */
+public class MlPipelineDockerPomMigration extends AbstractPomMigration {
+
+    private static final String PROJECT_GROUP_ID = "${project.groupId}";
+    private static final String STEP_PACKAGE_OUTPUT_DIR = "^<outputDirectory>\\$\\{project\\.build\\.directory}/(.+)<\\/outputDirectory>$";
+
+    @Override
+    protected boolean shouldExecuteOnFile(File file) {
+        Model model = PomHelper.getLocationAnnotatedModel(file);
+        if (model.getPackaging().equals("docker-build")) {
+            String stepArtifactId = getMlPipelineStepArtifact(file);
+            return stepArtifactId != null && !model.getDependencies().stream()
+                    .anyMatch(dependency -> isMlStep(dependency, stepArtifactId));
+        }
+        return false;
+    }
+
+    @Override
+    protected boolean performMigration(File file) {
+        Model model = PomHelper.getLocationAnnotatedModel(file);
+        String stepArtifactId = getMlPipelineStepArtifact(file);
+        detectAndSetIndent(file);
+        PomModifications modifications = new PomModifications();
+        // check if there is dependencies
+        if (model.getDependencies().size() > 0) {
+            String content = getDependency(stepArtifactId);
+            modifications.add(
+                    new PomModifications.Insertion(model.getLocation("dependencies" + END), 2, ignore -> content));
+        } else {
+            String content = getDependencies(stepArtifactId);
+            modifications.add(
+                    new PomModifications.Insertion(model.getLocation(END), 1, ignore -> content));
+        }
+
+        PomHelper.writeModifications(file, modifications.finalizeMods());
+        return true;
+    }
+
+    private static boolean isMlStep(Dependency dependency, String expectedArtifact) {
+        return dependency != null &&
+                dependency.getGroupId().equals(PROJECT_GROUP_ID) &&
+                dependency.getArtifactId().equals(expectedArtifact);
+    }
+
+    private String getMlPipelineStepArtifact(File file) {
+        try {
+            String step = null;
+            List<String> lines = Files.readAllLines(file.toPath());
+            for (String line: lines) {
+                line = line.trim();
+                if (line.matches(STEP_PACKAGE_OUTPUT_DIR)) {
```

Comment:
S: I'm a little worried about that there may be other lines besides the pipeline step copy that match, or that the pipeline step copy was modified. What if we instead checked to see if any of the collected ML steps are referenced? Something like:
```java
stepArtifactIds = getStepArtifactIds();
if (line.matches("<outputDir>.*<outputDir>") {
  var o = stepArtifactIds.stream().filter(line::contains).findFirst();
  if(o.isPresent()) {
    return o.get();
  }
}
```